src/game_events/conditional_wml.cpp

```diff
@@ -48,6 +49,14 @@ namespace game_events {
 namespace builtin_conditions {
 	std::vector<std::pair<int,int>> default_counts = utils::parse_ranges("1-infinity");
 
+	bool has_achievement(const vconfig& cfg)
+	{
+		const config& acfg = cfg.get_config();
+		std::string content_for = acfg["content_for"].str();
+		std::string id = acfg["id"].str();
+		return preferences::achieved(content_for, id);
```

Comment:
This should probably have anti-OOS protection in network multiplayer. However, maybe it should still be active in local multiplayer.

Reply:
What do you mean by anti-OOS protection?

Reply:
Consider how these would work in co-op multiplayer mode:

* a bonus "make it a bit more difficult" mini-boss that only appears if the player has already completed the campaign
* a multiplayer co-op campaign where there's a shortcut in an early map that's unlocked if you've already completed the campaign
* a whose recruit list depends on which units they've killed in previous scenarios (like the leaders of SotA), but which persistently unlocks those recruits even when loaded in an unrelated match
* a unit with a complex special ability which pops up several dialogs to control what it actually does; and after complaints about players trying it and then quitting, gets changed to only unlock if the player has completed the scenario which explains it
* a unit with a special ability which, if used wrongly, can block progress for several turns, which unlocks when you've played a scenario that explains it

These can all already be done with persistent storage; however persistent storage already has warnings about multiplayer OOS and how to avoid it. If we're making an tag to read these values easily via ConditionalWML, what are you expecting it to be used for?

Reply:
I understand that different clients will likely have different achievements completed, and as such isn't safe to use in networked multiplayer. If that's something we don't want to let people use due to it being incorrectly used in networked multiplayer, then I can of course remove the tag.

Reply:
I think stevecotton's point is more that we should warn the addon developer if they're using it in a way that might be unsafe?

Reply:
Yes, warn the addon developer, assuming that if it's accessible as ConditionalWML then addon developers will cut & paste without noticing surrounding warnings. However, we can assume that MP developers will at least test their MP code once...

So an OOS error `if(resources::controller->is_networked_mp() && ! in mp_sync::get_user_choice() or similar)`.

Reply:
Alright, that makes more sense to me. It's not protection as in preventing it from happening, it's warning that it happened.

Reply:
This has been completed.